src/physics/systems.rs

```diff
@@ -95,39 +100,61 @@ pub fn step_world_system(
 
 /// System responsible for writing the rigid-bodies positions into the Bevy translation and rotation components.
 pub fn sync_transform_system(
+    sim_to_render_time: Res<SimulationToRenderTime>,
     bodies: ResMut<RigidBodySet>,
     configuration: Res<RapierConfiguration>,
     rigid_body: &RigidBodyHandleComponent,
     mut transform: Mut<Transform>,
 ) {
+    let dt = sim_to_render_time.diff;
     if let Some(rb) = bodies.get(rigid_body.handle()) {
-        let pos = rb.position;
-
         #[cfg(feature = "dim2")]
         {
-            let rot = na::UnitQuaternion::new(na::Vector3::z() * pos.rotation.angle());
+            // Predict position at render time
+            let pos_0 = Vec2::new(
+                rb.position.translation.vector.x,
+                rb.position.translation.vector.y,
+            );
+            let vel_0 = Vec2::new(rb.linvel.x, rb.linvel.y);
+            let pos_t = pos_0 + vel_0 * dt;
```

Comment:
This looks like extrapolation rather than interpolation, which could show an entity at a position it will never be at in case of sudden turns.

Reply:
True. So then input events between simulation steps need to be taken into account for the rendering.

I was thinking it would be useful to ask the physics engine to predict the position of the body at t seconds ahead of where the simulation is at.

Reply:
I wanted to take acceleration state into account for the extrapolation as well but those fields weren’t accessible.

Reply:
I looked at the gaffer on games article again and it seems they suggest interpolating between the previous physics state and the current based on the time remaining in the accumulator. This seems backward in time to me. We just simulated the current state based on input that happened before the current simulation time, whatever remains in the accumulator is time into the future as far as the simulation is concerned, but the article suggests interpolating between two past points in time. It seems it’s suggesting to introduce a time step of lag.

Reply:
This is now rewritten to do what the article suggested - store the previous position and rotation, lerp the position, and slerp the rotation. It introduces a physics time step of lag though. :/